### PR TITLE
Document how to write & test analyzers

### PR DESCRIPTION
**Pull request by sushicw**
Add a README.md that documents how to create & test analyzers. 

See https://github.com/istio/istio/issues/17315

Preview it here: https://github.com/sushicw/istio/blob/config-analysis-readme/galley/pkg/config/analysis/README.md

[X] Configuration Infrastructure
[X] Docs
[ ] Installation
[ ] Networking
[ ] Performance and Scalability
[ ] Policies and Telemetry
[ ] Security
[ ] Test and Release
[ ] User Experience
[ ] Developer Infrastructure

### CODE REVIEW

File: galley/pkg/config/analysis/README.md [hunk 1]
```diff
@@ -0,0 +1,189 @@
...
+Analyzers need to implement the Analyzer interface ( in the `galley/pkg/config/analysis` package). They should be created under the analyzers subdirectory, and given their own appropriate subpackage.
+
+An annotated example:
+
+```go
+package virtualservice
+
+// <imports here>
+
+type GatewayAnalyzer struct{}
```
Comment (xeviknal) [inline galley/pkg/config/analysis/README.md]: Question: does a GatewayAnalyzer analyze all the possible invalid scenarios or only the scenario where the input VS has a Gateway that doesn't exist in the cluster? 
In that case, it is just a NIT, a more specific name like `NoGatewayAnalyzer` would help devs not put more than one responsibility per analyzer. I can imagine that there could be multiple scenarios to check for the gateways field in a VS.
For example, on Kiali we have all the different checkers split by responsibility: [Virtual Service checkers](https://github.com/kiali/kiali/tree/master/business/checkers/virtual_services). This way we apply Open/Close Principle.

**Comment by xeviknal** [comment]
Perhaps is not the place to ask this, but Is it there the intention to add path to yaml? Path as the unequivocal representation of where exactly is the problem in the yaml file. (e.g. spec/http[1]/route[2]/destination/host).

Besides those two comments, LGTM. It seems a great starting point.

**Comment by burmanm** [comment]
The feature is missing from the actual implementation, not just documentation. The yaml addition would indicate it's always a static path, which it isn't necessarily. So I think the fix should be done to the metadata of errors.

**Comment by ozevren** [comment]
We discussed the path model in one of the design doc comment threads. I think that is a very reasonable thing to add to the core model.

File: galley/pkg/config/analysis/README.md (hunk 1 above)
Comment (sushicw) [reply galley/pkg/config/analysis/README.md]: The analyzer in this doc is just intended as an example (although we do currently have an actual analyzer that's basically identical.) That said, if there are specific best practices we want to agree on around naming & organization, this doc will be a good place to record them.

**Comment by sushicw** [comment]
Created https://github.com/istio/istio/issues/17349 to track the work of adding pathing support. I agree that this would be valuable.

**Comment by esnible** [comment]
Need a section on adding to _galley/pkg/config/analysis/msg/messages.yaml_ and regenerating _messages.gen.go_.

**Comment by sushicw** [comment]
> Need a section on adding to _galley/pkg/config/analysis/msg/messages.yaml_ and regenerating _messages.gen.go_.

@esnible Those sections both exist: https://github.com/istio/istio/blob/master/galley/pkg/config/analysis/README.md#3-create-new-message-types

**Pull request merged**